crates/storage/src/lazy/mapping.rs

```diff
@@ -85,14 +86,27 @@ where
     where
         Q: scale::EncodeLike<K>,
     {
-        pull_packed_root_opt(&self.storage_key(key))
+        pull_packed_root_opt(&self.storage_key(&key))
+    }
+
+    /// Clears the value at `key` from storage.
+    ///
+    /// Returns the value at the key if it previously existed, otherwise returns `None`.
+    pub fn remove<Q>(&self, key: Q) -> Option<V>
+    where
+        Q: scale::EncodeLike<K>,
+    {
+        let storage_key = self.storage_key(&key);
+        let v = self.get(key)?;
+        clear_packed_root(&v, &storage_key);
+        Some(v)
     }
 
     /// Returns a `Key` pointer used internally by the storage API.
     ///
     /// This key is a combination of the `Mapping`'s internal `offset_key`
     /// and the user provided `key`.
-    fn storage_key<Q>(&self, key: Q) -> Key
+    fn storage_key<Q>(&self, key: &Q) -> Key
```

Comment:
I dont think this change makes sense since `EncodeLike` already allows to take references.

Reply:
The compiler says otherwise 😛

Reply:
Then there might be some misunderstanding or something might be wrong that we need to inspect further:
https://docs.rs/parity-scale-codec/2.3.1/parity_scale_codec/trait.EncodeLike.html#impl-EncodeLike%3CT%3E-for-%26T

Reply:
So if I take out the reference from the method signature I get the following error:

```
error[E0277]: the trait bound `&Q: EncodeLike<K>` is not satisfied
  --> crates/storage/src/lazy/mapping.rs:99:44
   |
99 |         let storage_key = self.storage_key(&key);
   |                                ----------- ^^^^ the trait `EncodeLike<K>` is not implemented for `&Q`
   |                                |
   |                                required by a bound introduced by this call
   |
help: consider removing the leading `&`-reference
   |
99 -         let storage_key = self.storage_key(&key);
99 +         let storage_key = self.storage_key(key);
   |
help: consider extending the `where` bound, but there might be an alternative better way to express this requirement
   |
69 |     V: PackedLayout, &Q: EncodeLike<K>
   |                    ~~~~~~~~~~~~~~~~~~~

For more information about this error, try `rustc --explain E0277`.
```

> Then there might be some misunderstanding or something might be wrong that we need to inspect further:
https://docs.rs/parity-scale-codec/2.3.1/parity_scale_codec/trait.EncodeLike.html#impl-EncodeLike%3CT%3E-for-%26T

This doesn't mean that you can put a `&T` when a `T` is expected by a method signature.

```rust
use parity_scale_codec::{EncodeLike, Encode};

fn foo<T: Encode, U: EncodeLike<T>>(t: U) {
    unimplemented!()
}

fn bar<T: Encode, U: EncodeLike<T>>(t: &U) {
    unimplemented!()
}

fn test() {
    foo::<u8, u8>(1u8);
    foo::<u8, u8>(&1u8); // Doesn't compile

    bar::<u8, u8>(&1u8);
    bar::<u8, u8>(1u8); // Doesn't compile
}
```

There was also [a comment by Alex](https://github.com/paritytech/ink/pull/979/files#r743079204) suggesting we turn this into a reference anyways to alleviate some costs in Wasm.

Reply:
That's because you should not use `let storage_key = self.storage_key(&key);` but simply `let storage_key = self.storage_key(key);` since `key: Q` is already bound to be `EncodeLike`.

Reply:
Since I want to call `Mapping::get(key: Q)` after computing the storage key I can't do that since it would move `key`